base/binaryplatforms.jl

```diff
@@ -0,0 +1,999 @@
+            # Note that in our effort to be extremely compatible, we actually allow something that
+            # doesn't parse nicely into a VersionNumber to persist, but if `validate_strict` is
+            # set to `true`, it will cause an error later on.
+            if tag ∈ ("libgfortran_version", "libstdcxx_version", "os_version")
+                normver(x::VersionNumber) = string(x)
+                function normver(str::AbstractString)
+                    v = tryparse(VersionNumber, str)
+                    if v === nothing
+                        # If this couldn't be parsed as a VersionNumber, return the original.
+                        return str
+                    end
+                    # Otherwise, return the `string(VersionNumber(str))` version.
+                    return normver(v)
+                end
+                value = normver(value)
+            end
+
+            # I know we said only alphanumeric and dots, but let's be generous so that we can expand
+            # our support in the future while remaining as backwards-compatible as possible.  The
+            # only characters that are absolutely disallowed right now are `-`, `+`, ` ` and things
+            # that are illegal in filenames:
+            nonos = raw"""+- /<>:"'\|?*"""
+            if any(occursin(nono, tag) for nono in nonos)
+                throw(ArgumentError("Invalid character in tag name \"$(tag)\"!"))
+            end
+
+            # Normalize and reject nonos
+            value = lowercase(string(value))
+            if any(occursin(nono, value) for nono in nonos)
+                throw(ArgumentError("Invalid character in tag value \"$(value)\"!"))
+            end
+            tags[tag] = value
+        end
+
+        # Auto-map call_abi and libc where necessary:
+        if os == "linux" && !haskey(tags, "libc")
+            # Default to `glibc` on Linux
+            tags["libc"] = "glibc"
+        end
+        if os == "linux" && arch ∈ ("armv7l", "armv6l") && "call_abi" ∉ keys(tags)
+            # default `call_abi` to `eabihf` on 32-bit ARM
+            tags["call_abi"] = "eabihf"
+        end
+
+        # If the user is asking for strict validation, do so.
+        if validate_strict
+            validate_tags(tags)
+        end
+
+        # By default, we compare julia_version only against major and minor versions:
+        if haskey(tags, "julia_version") && !haskey(compare_strategies, "julia_version")
+            compare_strategies["julia_version"] = (a::String, b::String, a_comparator, b_comparator) -> begin
+                a = VersionNumber(a)
+                b = VersionNumber(b)
+                return a.major == b.major && a.minor == b.minor
+            end
+        end
+
+        return new(tags, compare_strategies)
+    end
+end
+
+# Other `Platform` types can override this (I'm looking at you, `AnyPlatform`)
+tags(p::Platform) = p.tags
```

Comment:
Might want to use @timholy s "interface" trick like in 

https://github.com/JuliaLang/julia/commit/48fc8b4e5312eb9d815c7bfe665c55625a225654

here. Basically

```jl
function Base.getproperty(platform::AbstractPlatform, fieldname::Symbol)
    if fieldname === :tags
        return getfield(platform, :state)::Dict{String, String}
    else
        getfield(platform, fieldname)
    end
end
```

and then use `.tags` everywhere. That ensures inference know that `.tags` has to be a `Dict{String, String}`.

Reply:
I thought this kind of overriding of `getproperty()` was frowned upon?  Is that an outdated Julia style?

Reply:
This isn't about adding OOP-like behavior; it's really about fixing inference problems you have abstract types but can't infer the specific concrete subtype. (E.g., if you extract `dict[key]` when `dict::Dict{String,Platform}`, you know the result is a `Platform` but not the concrete subtype.) Inference will generally return `::Any` as the inferred type of field accesses to abstract types. Defining a `getproperty` lets you add the type-asserts to fix it. See https://timholy.github.io/SnoopCompile.jl/stable/snoopr/#Inferrable-field-access-for-abstract-types-1 for more explanation.

Reply:
I'm a little confused then because there are no abstract types here.  `Platform` is concrete, and `tags` is also concrete, it's a `Dict{String,String}`.  I don't want to define a `tags()` for other `AbstractPlatform` types because they will likely do something completely different (e.g. `AnyPlatform` won't have _any_ tags, it will just return a constant empty set, and other objects like the compatibility `Linux` type will sub out to a `tags()` on a wrapped `Platform` object).

Reply:
Then it doesn't matter. I was going by the current implementation for which
```julia
julia> isabstracttype(Pkg.BinaryPlatforms.Platform)
true
```

Reply:
>  I don't want to define a tags() for other AbstractPlatform types because they will likely do something completely different 

I assumed you would since there is a bunch of code that calls `tags` on `AbstractPlatform` e.g.

```
 Base.haskey(p::AbstractPlatform, k::String) = haskey(tags(p), k)
```

and

```
# Simple equality definition; for compatibility testing, use `platforms_match()`
Base.:(==)(a::AbstractPlatform, b::AbstractPlatform) = tags(a) == tags(b)
```

etc. Should these be restricted to `Platform` then?

Reply:
> Then it doesn't matter. I was going by the current implementation for which

Gotcha!  Makes sense.

> I assumed you would since there is a bunch of code that calls `tags` on AbstractPlatform

So the semantics of what `tags` does is stable; `tags(p::AbstractPlatform)` can be assumed to return a `Dict{String,String}`, but how it generates that dict is completely up to the `AbstractPlatform` object.  For an `AnyPlatform` object it will likely return an empty Dict.  For a `Linux` object (a compatibility wrapper for older BinaryProvider releases and such) it will sub out to an internal `Platform` object, etc...  So I still want to call `tags()` on other `AbstractPlatforms`, but I'm leaving it up to those other `AbstractPlatform` types to define just what needs to happen inside of `tags()`.